File: programmers/C_language/카드뭉치.c
#include <stdio.h>
#include <stdbool.h>
#include <stdlib.h>
#include <string.h>

// cards1_len은 배열 cards1의 길이입니다.
// cards2_len은 배열 cards2의 길이입니다.
// goal_len은 배열 goal의 길이입니다.
// 파라미터로 주어지는 문자열은 const로 주어집니다.
char* solution(
    const char* cards1[],
    size_t cards1_len,
    const char* cards2[],
    size_t cards2_len,
    const char* goal[],
    size_t goal_len
) {

    int idx1 = 0;
    int idx2 = 0;

    for(int i = 0; i < goal_len; i++) {

        // cards1에서 사용 가능한 경우
        if(idx1 < cards1_len &&
           strcmp(cards1[idx1], goal[i]) == 0) {

            idx1++;
        }

        // cards2에서 사용 가능한 경우
        else if(idx2 < cards2_len &&
                strcmp(cards2[idx2], goal[i]) == 0) {

            idx2++;
        }

        // 둘 다 불가능한 경우
        else {

            char* answer = (char*)malloc(3);

            strcpy(answer, "No");

            return answer;
        }
    }

    char* answer = (char*)malloc(4);

    strcpy(answer, "Yes");

    return answer;
}

File: Java/5월/05_30/카드 뭉치.java
class Solution {
    public String solution(String[] cards1, String[] cards2, String[] goal) {

      //원하는 카드뭉치 카드 순서대로 1장씩
      //한번 사용한 카드는 다시 사용 X
      //카드를 사용하지않고 다음카드 이동 X
      //기존에 주어진 카드뭉치의 단어순서 변경 X
      int idx1=0, idx2=0;

      //cards1, cards2를 둘러보면서 goal 단어 처리
      for(int i=0; i<goal.length; i++){

        //cards1에서 사용 가능한 경우
        if(idx1< cards1.length && cards1[idx1].equals(goal[i])){
          idx1++;
          continue;
        }

        //cards2에서 사용 가능한 경우
        if(idx2< cards2.length && cards2[idx2].equals(goal[i])){
          idx2++;
          continue;
        }

        //둘 다 사용 안되는 경우
        else {
          return "No";
        }
      }
      
      return "Yes";
    }
}